src/v/compression/async_stream_zstd.cc

```diff
@@ -0,0 +1,262 @@
+
+    // NOTE: always enable content size. **decompression** depends on this
+    throw_if_error(
+      ZSTD_CCtx_setPledgedSrcSize(ctx._compress_ctx, i_buf.size_bytes()));
+
+    iobuf ret_buf;
+
+    // if the buffer is empty encode and return an empty zstd frame.
+    if (i_buf.empty()) {
+        size_t remaining = 0;
+
+        do {
+            ZSTD_outBuffer out = {
+              .dst = ctx._c_buffer.get_write(),
+              .size = ctx._c_buffer.size(),
+              .pos = 0};
+            remaining = ZSTD_endStream(ctx._compress_ctx, &out);
+            throw_if_error(remaining);
+            ret_buf.append(ctx._c_buffer.get(), out.pos);
+        } while (remaining > 0);
+
+        co_return std::move(ret_buf);
+    }
+
+    // otherwise compress and return the buffer.
+    for (auto it = i_buf.cbegin(); it != i_buf.cend();) {
+        ZSTD_inBuffer in = {.src = it->get(), .size = it->size(), .pos = 0};
+        ++it;
+
+        const auto mode = (it == i_buf.cend()) ? ZSTD_e_end : ZSTD_e_continue;
+
+        int finished;
+        do {
+            ZSTD_outBuffer out = {
+              .dst = ctx._c_buffer.get_write(),
+              .size = ctx._c_buffer.size(),
+              .pos = 0};
+
+            const auto remaining = ZSTD_compressStream2(
+              ctx._compress_ctx, &out, &in, mode);
+            throw_if_error(remaining);
+
+            ret_buf.append(ctx._c_buffer.get(), out.pos);
+            finished = (mode == ZSTD_e_end) ? (remaining == 0)
+                                            : (in.pos == in.size);
+
+            co_await ss::coroutine::maybe_yield();
+        } while (!finished);
+    }
+
+    co_return std::move(ret_buf);
+}
+
+ss::future<iobuf> async_stream_zstd::uncompress(iobuf i_buf) {
+    if (unlikely(i_buf.empty())) {
+        throw std::runtime_error(
+          "Asked to stream_zstd::uncompress empty buffer");
+    }
+
+    ss::abort_source as;
+    auto ctx_scope = co_await _decompression_ctx_pool.allocate_scoped_object(
+      as);
+    auto& ctx = *ctx_scope;
+
+    ZSTD_DCtx_reset(ctx._decompress_ctx, ZSTD_reset_session_only);
+
+    iobuf ret_buf;
+    size_t last_zstd_ret = 0;
+
+    for (auto& ibuf : i_buf) {
+        ZSTD_inBuffer in = {.src = ibuf.get(), .size = ibuf.size(), .pos = 0};
+
+        while (in.pos < in.size) {
+            ZSTD_outBuffer out = {
+              .dst = ctx._d_buffer.get_write(),
+              .size = ctx._d_buffer.size(),
+              .pos = 0};
+
+            const auto zstd_ret = ZSTD_decompressStream(
+              ctx._decompress_ctx, &out, &in);
+            throw_if_error(zstd_ret);
+
+            ret_buf.append(ctx._d_buffer.get(), out.pos);
+            last_zstd_ret = zstd_ret;
+
+            co_await ss::coroutine::maybe_yield();
+        }
+    }
+
+    if (last_zstd_ret != 0) {
+        throw std::runtime_error("Input truncated before reading epilog");
+    }
+
+    co_return std::move(ret_buf);
```

Comment:
as above

Reply:
Removing the explicit move to allow for elision.